### ISSUE TITLE
Fatal Exception: NSInvalidArgumentException *** -[NSURL initFileURLWithPath:isDirectory:]: nil string parameter, +[ADJUtil writeObject:fileName:objectName:]

### ISSUE THREAD

**Issue by sspitzer** [adjust/ios_sdk#320]
Hi Adjust team,

We're seeing a new crash in +[ADJUtil writeObject:fileName:objectName:]

```
Fatal Exception: NSInvalidArgumentException *** -[NSURL initFileURLWithPath:isDirectory:]: nil string parameter
```

We recently upgraded from:

```
-  - Adjust (4.11.5):
 -    - Adjust/Core (= 4.11.5)
 -  - Adjust/Core (4.11.5)
 -  - Adjust/Criteo (4.11.5):
```

To:

```
 +  - Adjust (4.12.1):
 +    - Adjust/Core (= 4.12.1)
 +  - Adjust/Core (4.12.1)
 +  - Adjust/Criteo (4.12.1):
```

Here's the stack.  Let me know if you need more information:

```
Fatal Exception: NSInvalidArgumentException
0  CoreFoundation                 0x18531e364 __exceptionPreprocess
1  libobjc.A.dylib                0x184564528 objc_exception_throw
2  CoreFoundation                 0x18531e2ac -[NSException initWithCoder:]
3  Foundation                     0x185c2e268 -[NSURL(NSURL) initFileURLWithPath:isDirectory:]
4  Foundation                     0x185c2e108 +[NSURL(NSURL) fileURLWithPath:isDirectory:]
5  Foundation                     0x185cfa6bc _NSCreateTemporaryFile_Protected
6  Foundation                     0x185ce179c +[NSKeyedArchiver archiveRootObject:toFile:]
7  Adjust                         0x101153c40 +[ADJUtil writeObject:fileName:objectName:] (ADJUtil.m:382)
8  Adjust                         0x1011387f8 -[ADJActivityHandler writeActivityStateI:] (ADJActivityHandler.m:1308)
9  Adjust                         0x101135790 -[ADJActivityHandler processSessionI:] (ADJActivityHandler.m:707)
10 Adjust                         0x1011353e4 -[ADJActivityHandler startI:] (ADJActivityHandler.m:683)
11 Adjust                         0x101135294 -[ADJActivityHandler initI:preLaunchActionsArray:] (ADJActivityHandler.m:670)
12 Adjust                         0x1011326d0 __52-[ADJActivityHandler initWithConfig:savedPreLaunch:]_block_invoke_2 (ADJActivityHandler.m:195)
13 Adjust                         0x101157c20 __42+[ADJUtil launchInQueue:selfInject:block:]_block_invoke (ADJUtil.m:1221)
14 libdispatch.dylib              0x184c9aa54 _dispatch_call_block_and_release
15 libdispatch.dylib              0x184c9aa14 _dispatch_client_callout
16 libdispatch.dylib              0x184ca496c _dispatch_queue_serial_drain$VARIANT$mp
17 libdispatch.dylib              0x184ca52fc _dispatch_queue_invoke$VARIANT$mp
18 libdispatch.dylib              0x184ca5d20 _dispatch_root_queue_drain_deferred_wlh$VARIANT$mp
19 libdispatch.dylib              0x184cae03c _dispatch_workloop_worker_thread$VARIANT$mp
20 libsystem_pthread.dylib        0x184f42f1c _pthread_wqthread
21 libsystem_pthread.dylib        0x184f42b6c start_wqthread
```

**Comment by sspitzer** [comment]
We see it on iOS 11 and iOS 10 if that helps.

**Comment by uerceg** [comment]
Hi @sspitzer,

Thanks for the report. Is the issue reproducible and if yes, what are the steps?

**Comment by sspitzer** [comment]
We haven't not been able to reproduce it, we are just seeing it in fabric.

**Comment by sspitzer** [comment]
But multiple users are running into this.  We plan to go back to version 4.11.5

**Comment by uerceg** [comment]
Any info on frequency of occurence for your users that use app version with 4.12.1 inside?

**Comment by sspitzer** [comment]
We've had twelve users hit the crash a total of 65 times since Dec 29, 2017.

**Comment by sspitzer** [comment]
Here's the thread stack trace:

```
io.adjust.PackageQueue
0  libsystem_kernel.dylib         0x184e11014 write + 8
1  ShopStyle                      0x100da4034 CLSSDKFileLog + 4337451060
2  ShopStyle                      0x100dadbac CLSTerminateHandler() + 4337490860
3  libc++abi.dylib                0x18455454c std::__terminate(void (*)()) + 16
4  libc++abi.dylib                0x184553ea8 __cxxabiv1::exception_cleanup_func(_Unwind_Reason_Code, _Unwind_Exception*) + 134
5  libobjc.A.dylib                0x18456465c _objc_exception_destructor(void*) + 362
6  CoreFoundation                 0x18531e2ac -[NSException initWithCoder:] + 114
7  Foundation                     0x185c2e268 -[NSURL(NSURL) initFileURLWithPath:isDirectory:] + 132
8  Foundation                     0x185c2e108 +[NSURL(NSURL) fileURLWithPath:isDirectory:] + 56
9  Foundation                     0x185cfa6bc _NSCreateTemporaryFile_Protected + 472
10 Foundation                     0x185ce179c +[NSKeyedArchiver archiveRootObject:toFile:] + 68
11 Adjust                         0x101153c40 +[ADJUtil writeObject:fileName:objectName:] (ADJUtil.m:382)
12 Adjust                         0x10114a0e4 -[ADJPackageHandler writePackageQueueS:] (ADJPackageHandler.m:258)
13 Adjust                         0x101149810 -[ADJPackageHandler addI:package:] (ADJPackageHandler.m:173)
14 Adjust                         0x101157c20 __42+[ADJUtil launchInQueue:selfInject:block:]_block_invoke (ADJUtil.m:1221)
15 libdispatch.dylib              0x184c9aa54 _dispatch_call_block_and_release + 24
16 libdispatch.dylib              0x184c9aa14 _dispatch_client_callout + 16
17 libdispatch.dylib              0x184ca496c _dispatch_queue_serial_drain$VARIANT$mp + 528
18 libdispatch.dylib              0x184ca52fc _dispatch_queue_invoke$VARIANT$mp + 340
19 libdispatch.dylib              0x184ca5d20 _dispatch_root_queue_drain_deferred_wlh$VARIANT$mp + 404
20 libdispatch.dylib              0x184cae03c _dispatch_workloop_worker_thread$VARIANT$mp + 644
21 libsystem_pthread.dylib        0x184f42f1c _pthread_wqthread + 932
22 libsystem_pthread.dylib        0x184f42b6c start_wqthread + 4
```

**Comment by sspitzer** [comment]
here's the full report:
[com.sugarinc.shopstyle_issue_558_crash_976d7654eed54cfbb2a7f8ac7108d538_b3114643f4ae11e78f7356847afe9799_0_v2.txt](https://github.com/adjust/ios_sdk/files/1613115/com.sugarinc.shopstyle_issue_558_crash_976d7654eed54cfbb2a7f8ac7108d538_b3114643f4ae11e78f7356847afe9799_0_v2.txt)

**Comment by uerceg** [comment]
Thanks for the report, we'll take a look and ping you in here as soon as we can.

**Comment by sspitzer** [comment]
great, thank you!

**Comment by uerceg** [comment]
@sspitzer Also headsup regarding downgrading Adjust SDK - I would advise you against this, especially having in mind that v4.12.1 includes long time requested features to move Adjust internal files from `Documents` folder into `Application Support` folder (for reference: https://github.com/adjust/ios_sdk/issues/171). Migration, like described in the issue was done for security reasons since with newer iOS versions, these files were actually quite easy accessible by users and potentially deleted.

So all internal SDK files were migrated to `Application Support/Adjust` directory to be hidden from users and to prevent it's potential deletion. If you revert back, our SDK will look for these files in old, `Documents, folder and if not found in there (it won't find them if migration has been done), it will create new ones and work with them.

In case some of your users had events buffered at new location and not sent, with downgrade you won't get them. Good thing is that files will remain in there (on new location), so upon potentially new switch from pre 4.12.0 to 4.12.0++, our SDK will read them at new location and send what ever was to be sent, but won't do migration, since it will think that files already exist at new location, so what ever got queued in downgraded version might get lost.

In theory, maybe everything will go smoothly, depends really on use cases in your app, but as always, we generally don't advise downgrade and this feature which 4.12.0 brought makes downgrade even tricker to version prior to that one.

Will update in here as soon as we have some findings about the crash.

**Comment by sspitzer** [comment]
Thanks for the warning about downgrading.

**Comment by uerceg** [comment]
Just as an info for now (since it's kinda late in Berlin), I didn't manage to reproduce this crash. Usually it is quite tricky when we're just unable to reproduce the crash, but then I tried to analyse crash stack trace. Going all the way up in calling hierarchy, last thing mentioned from Adjust SDK is this line:

```
[ADJUtil writeObject:fileName:objectName:] (ADJUtil.m:382)
```

That line in particular in v4.12.0 that you are using is this one: https://github.com/adjust/ios_sdk/blob/master/Adjust/ADJUtil.m#L382

Throughout the years I remember numerous issues with (at the beginning) lack of line numbering in crash reports but also (once finally made available), "wrong" line numbering in stack traces. This line in particular shows the line where we're doing simple `if` statement, which probably has nothing to do with the crash (also, following stack trace all the way up showed lack of accuracy in line numbers, but if simply methods existing inside of scope aimed by number are followed, trace can be recreated, but not by following line numbers from the trace).

Next line in call mentions:

```
[NSKeyedArchiver archiveRootObject:toFile:] + 68
```

Which if we look at method whose scope exists in mentioned line 382 is probably caused by call in line 380: https://github.com/adjust/ios_sdk/blob/master/Adjust/ADJUtil.m#L380

Later calls are made within `Foundation`/`CoreFoundation` iOS frameworks:

```
2  CoreFoundation                 0x18531e2ac -[NSException initWithCoder:]
3  Foundation                     0x185c2e268 -[NSURL(NSURL) initFileURLWithPath:isDirectory:]
4  Foundation                     0x185c2e108 +[NSURL(NSURL) fileURLWithPath:isDirectory:]
5  Foundation                     0x185cfa6bc _NSCreateTemporaryFile_Protected
```

If we assume that line 380 is the last call made inside of Adjust SDK that lead to the crash, we're talking about following line:

```objc
BOOL result = [NSKeyedArchiver archiveRootObject:object toFile:filePath];
```

Stack trace from iOS frameworks tells that issue was happening while path towards destination file, in which object is supposed to be archived, was being created with usage of `NSURL` class. Two methods from the stack trace are: https://developer.apple.com/documentation/foundation/nsurl/1417505-initfileurlwithpath and https://developer.apple.com/documentation/foundation/nsurl/1414650-fileurlwithpath. Documentation doesn't mention anything about potential crashes they might cause.

I did not manage to find out in the documentation the logic that causes `fileUrlWithPath:isDirectory:` to be invoked after `archiveRootObject:toFile:` is called. In my test I have tested archiving serialisable and also not serialisable and `nil` object into path that `exists`, `doesn't exit`, `is nil`. I didn't manage to cause exception to be thrown in any of those situations.

What I managed to do is to explicitly make a call to `fileURLWithPath:isDirectory:` method on my own by passing `nil` as a path:

```objc
[NSURL fileURLWithPath:nil isDirectory:NO]; // or YES, same thing
```

With this, I managed to get exception being thrown, but still not really looking identically to stack trace from your crash example:

```
	0   CoreFoundation                      0x00000001037ac12b __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x0000000102e40f41 objc_exception_throw + 48
	2   CoreFoundation                      0x0000000103821245 +[NSException raise:format:] + 197
	3   Foundation                          0x0000000102831d0e -[NSURL(NSURL) initFileURLWithPath:isDirectory:] + 149
	4   Foundation                          0x0000000102831bc5 +[NSURL(NSURL) fileURLWithPath:isDirectory:] + 53
```

So, if somehow underlying calls to `Foundation` methods manage to pass `nil` as path value to `fileURLWithPath:isDirectory:` method, seems like this exception might happen. I am still not sure how can this happen with making a call to `archiveRootObject:toFile:`.

Also, Adjust SDK isn't making a call to `fileURLWithPath:isDirectory:` at all. But if we move one line below (383), inside of call to `excludeFromBackup:` method, we do make a call to `fileURLWithPath:` method: https://github.com/adjust/ios_sdk/blob/master/Adjust/ADJUtil.m#L204 which, if invoked with `nil` might cause the same exception from above (the one I managed to recreate). But then again, stack trace from your crash log doesn't show call to `excludeFromBackup:` being made at all, so probably this code path doesn't cause the issue.

So, to sum up - I am not really sure how iOS is internally chaining these calls after call to `archiveRootObject:toFile:`, nor how underlying `Foundation` methods can get `nil` as path parameter. From all these things I managed to explore, the only thing that managed to force exception being thrown in certain artificial calls I made was passing `nil` as `path` value.

The only thing that comes up to my mind currently can be that after we obtain `filePath` value in this line: https://github.com/adjust/ios_sdk/blob/master/Adjust/ADJUtil.m#L378, we abort further method execution if path is `nil`. Even though this value might be returned as `nil`, according to tests we made, this line https://github.com/adjust/ios_sdk/blob/master/Adjust/ADJUtil.m#L380 should handle that w/o any problems and return `NO` which will prevent further logic from being executed. But maybe as even sooner validation, we can add this `nil` check. But like said above, I am not able to reproduce this issue and determine 100% what is causing this issue, so even if we go with this fix attempt (which if added I think it's still semantically correct), I don't think we can guarantee that it will solve these random crashes you spotted.

If you maybe have something to add to this or maybe you noticed something I didn't, feel free to ping in here.

**Comment by sspitzer** [comment]
Thanks for the detailed analysis!  

I like your idea of adding that nil check, but I agree that it might not fix the root of this problem.

Question regarding "BOOL result = [NSKeyedArchiver archiveRootObject:object toFile:filePath];" not having problems when you passed in nil.

Was that in Simulator or on a device?

Given your warning, we won't downgrade.  But we are willing to try a fix in the next version of our app if you have something.

**Comment by uerceg** [comment]
I tried that with different iOS versions on Simulator, correct. In several hours I will have access to iOS devices with different OS versions and was planning to redo the tests with them as well.

**Comment by uerceg** [comment]
Thanks for this observation, indeed like you probably had in mind, when testing on device, passing `nil` path to `archiveRootObject:toFile:` method call causes `NSInvalidArgumentException`.

Terminating message is following:

```
Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[NSURL initFileURLWithPath:isDirectory:]: nil string parameter
```

And stack trace looks like this:

![screen shot 2018-01-09 at 08 58 44](https://user-images.githubusercontent.com/1741635/34710431-9b5e620e-f51b-11e7-8858-d76f9cac51d6.png)

Comparing exception message with yours from above seems like that's the one. Also, by looking at device crash log, this is the thread that caused it:

```
Thread 3 Crashed:
0   libsystem_kernel.dylib        	0x00000001818e42e8 __pthread_kill + 8
1   libsystem_pthread.dylib       	0x00000001819f92f8 pthread_kill$VARIANT$mp + 396
2   libsystem_c.dylib             	0x0000000181852fbc abort + 140
3   libc++abi.dylib               	0x0000000180ff0068 __cxa_bad_cast + 0
4   libc++abi.dylib               	0x0000000180ff0210 default_unexpected_handler+ 8720 () + 0
5   libobjc.A.dylib               	0x0000000181018810 _objc_terminate+ 34832 () + 124
6   libc++abi.dylib               	0x000000018100854c std::__terminate(void (*)+ 107852 ()) + 16
7   libc++abi.dylib               	0x0000000181007ea8 __cxxabiv1::exception_cleanup_func+ 106152 (_Unwind_Reason_Code, _Unwind_Exception*) + 0
8   libobjc.A.dylib               	0x000000018101865c _objc_exception_destructor+ 34396 (void*) + 0
9   CoreFoundation                	0x0000000181dcf0ac -[NSException initWithCoder:] + 0
10  Foundation                    	0x00000001826de268 -[NSURL+ 135784 (NSURL) initFileURLWithPath:isDirectory:] + 132
11  Foundation                    	0x00000001826de108 +[NSURL+ 135432 (NSURL) fileURLWithPath:isDirectory:] + 56
12  Foundation                    	0x00000001827aa6bc _NSCreateTemporaryFile_Protected + 472
13  Foundation                    	0x000000018279179c +[NSKeyedArchiver archiveRootObject:toFile:] + 68
14  AdjustExample-iOS             	0x0000000100723e7c 0x10070c000 + 97916
15  libdispatch.dylib             	0x000000018174ea54 _dispatch_call_block_and_release + 24
16  libdispatch.dylib             	0x000000018174ea14 _dispatch_client_callout + 16
17  libdispatch.dylib             	0x0000000181755bc8 _dispatch_queue_override_invoke$VARIANT$mp + 716
18  libdispatch.dylib             	0x000000018175bcf4 _dispatch_root_queue_drain + 600
19  libdispatch.dylib             	0x000000018175ba38 _dispatch_worker_thread3 + 120
20  libsystem_pthread.dylib       	0x00000001819f706c _pthread_wqthread + 1268
21  libsystem_pthread.dylib       	0x00000001819f6b6c start_wqthread + 4
```

Crashing sequence from call to `archiveRootObject:toFile:` seems to match sequence from thread 1 that caused crash in the log you provided.

Now that this is was managed to be reproduced on the device, I am more convinced that `nil` path sent to `archiveRootObject:toFile:` might be causing these random crashes you are noticing. Why are those devices so interesting that path returned for them is `nil`, that's good question, but regardless of the cause, adding the check I mentioned should hopefully prevent these things from happening.

We will add this check into our next patch version and will update you in here once released.

In the meantime, if you have any other questions and if we can help you with anything else regarding this topic, feel free to ping in here.

**Comment by sspitzer** [comment]
Excellent work, thanks for digging deeper!  I'm also wondering why:

```
NSString *filePath = [ADJUtil getFilePathInAppSupportDir:fileName];
```

returns nil and what other problems that will cause for these users.

**Comment by sspitzer** [comment]
I forced getFilePathInAppSupportDir to always return nil and ran on my device and so far, did not see any other crashers (other than this one.)

My local "fix" is to do:

```
BOOL result = (filePath != nil) && [NSKeyedArchiver archiveRootObject:object toFile:filePath];
```

Let me know what you think.

**Comment by uerceg** [comment]
Lemme elaborate first a bit on other problems topic.

There are two possible scenarios in which `nil` will be returned: https://github.com/adjust/ios_sdk/blob/master/Adjust/ADJUtil.m#L430 and https://github.com/adjust/ios_sdk/blob/master/Adjust/ADJUtil.m#L436

Both moments pretty much mean -> new destination path to which want to migrate Adjust files did not manage to be created on these devices. Why - good question. 

Before, we had internal Adjust files located inside of `Documents` folder which, like described in linked issue from above, started to be easy for users to access it and actually delete stuff from it. So in accordance to Apple guidelines, we have moved our files into `Application Support` directory of the app bundle, additionally (not to pile up our files in root of that folder) we are creating folder called `Adjust` so all of our internal files are migrated to `Application Support/Adjust` folder inside of app bundle. Migration was tested for all supported iOS versions that our SDK supports (iOS 6 until 11) and was being performed successfully and without any issues. When SDK 4.12.0 initially starts, we are making sure that this migration is done properly and only after making sure that files were successfully moved to new location, we are removing files from `Documents` folder. Writing new information to those files is as of that moment assumed to be done to new location. That is logic behind the scenes.

Now for some reason, these couple of devices are constantly failing in following operation sequence:

- Check if this app bundle already has `Application Support` folder created. If no, create it.
- Check if this app bundle already has `Adjust` folder created inside of `Application Support` folder. If no, create it.

I am not sure what can possibly be the reason since I never manage to reproduce this. We do have logging in place for this scenario (https://github.com/adjust/ios_sdk/blob/master/Adjust/ADJUtil.m#L456), but logs are not being displayed in production environment and you probably have no way to access to this information.

Lack of device's ability to create and persist Adjust internal files pretty much means that persistence mechanisms SDK has thanks to these files are not present. This has nothing much to do with files being in new location, I can imagine certain devices for whatever reasons (maybe lack of available space or maybe some Jailbreak magic) not being able to persist things inside of app bundle and with that behaviour, of course, problem would exist regardless of where internal files are stored. Everything that SDK is attempting to track is persisted in memory and additionally inside of these files. So if internet access exists on device when whatever is attempted to be tracked (session/event), it will be tracked since in-memory queueing mechanism takes care of it, but if for some reason before network request for sending is made app crashes, due to lack of that package persistence into internal files, next time app starts, sending of that package will not be retried.

But this concern is nothing new or introduced with 4.12.0 and file migration, it's something that exist since beginning of SDK time and is simply like that.

**Comment by uerceg** [comment]
As far as your local fix is concerned, we might do it a bit differently syntax wise, but semantically, it will be identical to what your change is accomplishing. 👍

**Comment by sspitzer** [comment]
Thanks again for the detailed info!  We'll be shipping a new version of our app with that change and I'll let you know.  When the real fix is in your next update, we'll update to that version.  Any ETA?

**Comment by uerceg** [comment]
Currently no specific ETA. But having in mind that this change is not really big one, we might as well be able to push this update soon-ish. I don't really wanna promise something which turns out to be missed ETA, but I sincerely hope that we'll have it live by the end of next week (maybe even sooner, but like said, no promises).

**Comment by sspitzer** [comment]
Sounds good, thanks again!

**Comment by sspitzer** [comment]
@uerceg we've released a new version of our app with that local fix, and we're watching fabric for any new crashers.  I'll let you know how it goes.

**Comment by uerceg** [comment]
@sspitzer Many thanks for the update. We are also working on iOS SDK update that will contain this fix as well. Looking forward to hear if you noticed any crashes like this with your latest app update (hopefully not). :)

**Comment by sspitzer** [comment]
This fixed our crash and we have not noticed any other crashes in adjust.

**Comment by uerceg** [comment]
@sspitzer

Awesome, thanks for letting us know. Fix is already added to our code (locally), still prepairing some additional things we wanted to bundle in next release. Will ping in here once live. 

Thank you very much for time and willigness to test it on your own. 

Cheers

**Comment by VPTeamMobile** [comment]
Hi @uerceg , we also have this issue in Fabric. We have around 50 craches by day.
Do you know when will be the next release ?

**Comment by uerceg** [comment]
Hi @VPTeamMobile

4.12.2 is aimed to be live this week. Will keep you posted in here. Thanks for reporting!

**Comment by VPTeamMobile** [comment]
Hi @uerceg,

Any news on the release? 🙏 

Thanks

**Comment by uerceg** [comment]
@VPTeamMobile (and @sspitzer as well),

Yes. :rocket: https://github.com/adjust/ios_sdk/releases/tag/v4.12.2

Thank you one more time and feel free to reopen the issue if it still persists with this or later SDK versions.

Cheers. 🍺

**Comment by picturehit** [comment]
Thank you @uerceg I just updated the pod, you can close the ticket, i will reopen it if I see the bug again. 
Have a good day !
